aioxmpp/blocking/service.py

```diff
@@ -0,0 +1,225 @@
+########################################################################
+# File name: service.py
+# This file is part of: aioxmpp
+#
+# LICENSE
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program.  If not, see
+# <http://www.gnu.org/licenses/>.
+#
+########################################################################
+import asyncio
+
+import aioxmpp
+import aioxmpp.callbacks as callbacks
+import aioxmpp.service as service
+
+from aioxmpp.utils import namespaces
+
+from . import xso as blocking_xso
+
+
+class BlockingClient(service.Service):
+    """
+    A :class:`~aioxmpp.service.Service` implementing :xep:`Blocking
+    Command <191>`.
+
+    This service maintains the list of blocked JIDs and allows
+    manipulating the blocklist.
+
+    Attribute:
+
+    .. autoattribute:: blocklist
+
+    Signals:
+
+    .. signal:: on_initial_blocklist_reveived(blocklist)
+
+       Fires when the initial blocklist was received from the server.
+
+       :param blocklist: the initial blocklist
+       :type blocklist: :class:`~collections.abc.Set` of :class:`~aioxmpp.JID`
+
+    .. signal:: on_jids_blocked(blocked_jids)
+
+       Fires when additional JIDs are blocked.
+
+       :param blocked_jids: the newly blocked JIDs
+       :type blocklist: :class:`~collections.abc.Set` of :class:`~aioxmpp.JID`
+
+    .. signal:: on_jids_blocked(blocked_jids)
+
+       Fires when JIDs are unblocked.
+
+       :param unblocked_jids: the now unblocked JIDs
+       :type blocklist: :class:`~collections.abc.Set` of :class:`~aioxmpp.JID`
+
+    Coroutine methods:
+
+    .. automethod:: block_jids
+
+    .. automethod:: unblock_jids
+
+    .. automethod:: unblock_all
+    """
+    ORDER_AFTER = [aioxmpp.DiscoClient]
+
+    def __init__(self, client, **kwargs):
+        super().__init__(client, **kwargs)
+        self._blocklist = None
+        self._lock = asyncio.Lock()
+        self._disco = self.dependencies[aioxmpp.DiscoClient]
+
+        self._bse_token = client.before_stream_established.connect(
```

Comment:
Uh. We need depsignal to work on Client too, I see. Will do that at some point and then we can fix this, that’s out of scope for this PR.